nox/tasks.py

```diff
@@ -51,15 +52,26 @@ def load_nox_module(global_config: Namespace) -> Union[types.ModuleType, int]:
             os.path.expandvars(global_config.noxfile)
         )
 
+        # Check ``nox.needs_version`` by parsing the AST.
+        check_nox_version(global_config.noxfile)
+
         # Move to the path where the Noxfile is.
         # This will ensure that the Noxfile's path is on sys.path, and that
         # import-time path resolutions work the way the Noxfile author would
         # guess.
         os.chdir(os.path.realpath(os.path.dirname(global_config.noxfile)))
-        return importlib.machinery.SourceFileLoader(
+        module = importlib.machinery.SourceFileLoader(
             "user_nox_module", global_config.noxfile
         ).load_module()  # type: ignore
 
+        # Check ``nox.needs_version`` as set by the Noxfile.
+        check_nox_version()
```

Comment:
Hmm, why is this done twice? When would this return something different from the AST parsing version?

Reply:
Anytime the user does something more complex than assigning a string literal to `nox.needs_version`.

Here are some examples:

```
NOX_VERSION = ">=2022.1.1"
nox.needs_version = NOX_VERSION
```

```
import nox as _nox
_nox.needs_version = ">=2022.1.1"
```

```
nox.needs_version = ">=2022.1.1"

if sys.platform == "win32":
    nox.needs_version = None
```

The last example shows that the AST-based version check could actually lead to a
false positive: It would return `">=2022.1.1"`, ignoring the assignment in the
`if` clause. On Windows, Nox older than 2022.1.1 would therefore bail out early,
before even importing the Noxfile. It would never see the `if` clause that
permits any version on this platform.

I'm not sure how problematic this is, what do you think? For now, the
documentation suggests to assign a simple string literal, and I expect this is
what most users would do anyway.

Reply:
I should maybe elaborate more on why we would bother to parse the AST instead of
just importing the Noxfile. The idea is to avoid executing user code that wasn't
written for our version (as far as we can detect this). #210 took a different
approach here: Import the Noxfile, and only parse the source code if the import
failed. I think both approaches have their merits and drawbacks, and I'm happy
to take this either way.